cmd/speaker/app/speaker.go

```diff
@@ -83,78 +79,58 @@ func NewOpenELBSpeakerCommand() *cobra.Command {
 }
 
 func Run(c *options.OpenELBSpeakerOptions) error {
-	log.InitLog(c.LogOptions)
-
-	setupLog := ctrl.Log.WithName("speaker")
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&c.LogOptions.Options)))
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		// MetricsBindAddress: c.MetricsAddr,
+		Scheme: scheme,
+	})
 
-	mgr, err := manager.NewManager(ctrl.GetConfigOrDie(), c.GenericOptions)
-	setupLog.Info("listen metrics addr : " + c.MetricsAddr)
 	if err != nil {
 		setupLog.Error(err, "unable to new manager")
 		return err
 	}
 
+	//For gobgp
 	bgpServer := bgpd.NewGoBgpd(c.Bgp)
-
-	err = bgp.SetupBgpConfReconciler(bgpServer, mgr)
-	if err != nil {
+	if err := bgp.SetupBgpConfReconciler(bgpServer, mgr); err != nil {
 		setupLog.Error(err, "unable to setup bgpconf")
+		return err
 	}
 
-	err = bgp.SetupBgpPeerReconciler(bgpServer, mgr)
-	if err != nil {
+	if err := bgp.SetupBgpPeerReconciler(bgpServer, mgr); err != nil {
 		setupLog.Error(err, "unable to setup bgppeer")
-	}
-
-	if err = lb.SetupServiceReconciler(mgr); err != nil {
-		setupLog.Error(err, "unable to setup lb controller")
 		return err
 	}
 
-	stopCh := ctrl.SetupSignalHandler()
-
-	//For layer2
-	k8sClient := clientset.NewForConfigOrDie(ctrl.GetConfigOrDie())
-	leader.LeaderElector(stopCh.Done(), k8sClient, *c.Leader)
-
-	//For gobgp
-	err = speaker.RegisterSpeaker(constant.OpenELBProtocolBGP, bgpServer)
-	if err != nil {
+	// TODO: for layer2 + vip mode
+	spmanager := speaker.NewSpeakerManager(mgr.GetClient(), ctrl.Log.WithName("speakerManger"))
+	if err := spmanager.RegisterSpeaker(constant.OpenELBProtocolBGP, bgpServer); err != nil {
 		setupLog.Error(err, "unable to register bgp speaker")
 		return err
 	}
-	keepalive := vip.NewKeepAlived(k8sClient, &vip.KeepAlivedConfig{
-		Args: []string{fmt.Sprintf("--services-configmap=%s/%s", util.EnvNamespace(), constant.OpenELBVipConfigMap)},
-	})
 
-	//For keepalive
-	err = speaker.RegisterSpeaker(constant.OpenELBProtocolVip, keepalive)
-	if err != nil {
-		setupLog.Error(err, "unable to register keepalive speaker")
+	if err := (&speaker.LBReconciler{
+		Handler:       spmanager.HandleService,
```

Comment:
Why not use the `SetupLBReconciler` function?

Reply:
> Why not use the `SetupLBReconciler` function?

Reduced some code size and function calls :) ,nothing else.